File: ObjC-JSON-API/ObjC-JSON-API/Net/L3SDKJWARequest.m
//
//  JWARequest.m
//  ObjC-JSON-API
//
//  Created by Domenico Vacchiano on 07/11/15.
//  Copyright © 2015 DomenicoVacchiano. All rights reserved.
//

#import "L3SDKJWARequest.h"

@interface L3SDKJWARequest ()
+(NSString*)httpRequestMethodToString:(HTTPRequestMethod)httpRequestMethod;
@end


@implementation L3SDKJWARequest

+(void)sendTo:(NSString*)url
   withMethod:(HTTPRequestMethod)httpMethod
    andParams:(NSDictionary*)params
  andDelegate:(id<L3SDKJWARequestDelegate>)delegate{
    

    //creates and setup a request
    NSMutableURLRequest*request=[[NSMutableURLRequest alloc]
                                 initWithURL:[NSURL URLWithString:url]
                                 ];
    
    NSURLSession*session=[NSURLSession sharedSession];
    request.HTTPMethod=[self httpRequestMethodToString:httpMethod];
    if (params) {
        NSError *parseError = nil;
        request.HTTPBody=[NSJSONSerialization dataWithJSONObject:params options:NSJSONWritingPrettyPrinted error:&parseError];
        if (parseError) {
            if ([delegate respondsToSelector:@selector(L3SDKJWARequestDelegate_Error:)]) {
                [delegate L3SDKJWARequestDelegate_Error:parseError];
            }
            return;
        }
    }
    [request addValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    [request addValue:@"application/json" forHTTPHeaderField:@"Accept"];
    
    //sends a request to consume a service
    [[session dataTaskWithRequest:request completionHandler: ^(NSData * __nullable data, NSURLResponse * __nullable response, NSError * __nullable taskError){
        if (taskError) {
            if ([delegate respondsToSelector:@selector(L3SDKJWARequestDelegate_Error:)]) {
                [delegate L3SDKJWARequestDelegate_Error:taskError];
            }
            return;
        }
        NSError *parseError = nil;
        id json=[NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableLeaves error:&parseError];
        if (parseError) {
            if ([delegate respondsToSelector:@selector(L3SDKJWARequestDelegate_Error:)]) {
                [delegate L3SDKJWARequestDelegate_Error:parseError];
            }
        }else{
            if ([delegate respondsToSelector:@selector(L3SDKJWARequestDelegate_Response:)]) {
                [delegate L3SDKJWARequestDelegate_Response:json];
            }
        }
    }]resume];
  
}
+(NSString*)httpRequestMethodToString:(HTTPRequestMethod)httpRequestMethod{
    
    //convert an enum member to a string http verb
    switch (httpRequestMethod) {
        case HTTPRequestMethodGet:
            return @"GET";
        case HTTPRequestMethodPost:
            return @"POST";
        default:
            return @"GET";
            break;
    }
}
@end

File: Swift-JSON-API/Swift-JSON-API/Net/ L3SDKJWARequest.swift
//
//  L3SDKJWARequest.swift
//  Swift-JSON-API
//
//  Created by Domenico Vacchiano on 07/11/15.
//  Copyright © 2015 DomenicoVacchiano. All rights reserved.
//

import UIKit

class  L3SDKJWARequest: NSObject {
    
    //enumeration to describe the http verbs
    enum HTTPRequestMethod:String{
        case GET = "GET"
        case POST = "POST"
    }
   
    //method used to send a request to the service
    static func sendTo (url url:String,withMethod httpMethod:HTTPRequestMethod,andParams params:Dictionary<String,AnyObject>?,andDelegate delegate:L3SDKJWARequestDelegate){

        //creates and setup a request
        let request = NSMutableURLRequest(URL: NSURL(string: url)!)
        
        let session = NSURLSession.sharedSession()
        request.HTTPMethod = httpMethod.rawValue
        if (params != nil){
            request.HTTPBody = try! NSJSONSerialization.dataWithJSONObject(params!, options: [])
        }
        
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue("application/json", forHTTPHeaderField: "Accept")
        
        //sends a request to consume a service
        let task = session.dataTaskWithRequest(request, completionHandler: {data, response, taskError -> Void in
            
            if (taskError != nil) {
                delegate.L3SDKJWARequestDelegate_Error!(taskError!)
                return
            }
            
            do {
                let json = try NSJSONSerialization.JSONObjectWithData(data!, options: .MutableLeaves )
                delegate.L3SDKJWARequestDelegate_Response(json)
                
            } catch  let error as NSError  {
               //print(error.localizedDescription)
                delegate.L3SDKJWARequestDelegate_Error!(error)
            }
        
        })
    
        task.resume()
        
    }

}
